openedx/features/content_tagging/tasks.py

```diff
@@ -0,0 +1,41 @@
+"""
+Defines asynchronous celery task for auto-tagging content
+"""
+
+import logging
+from celery import shared_task
+from celery_utils.logged_task import LoggedTask
+from edx_django_utils.monitoring import set_code_owner_attribute
+from openedx_tagging.core.tagging.models import Taxonomy
+from opaque_keys.edx.keys import CourseKey, UsageKey
+from opaque_keys.edx.locator import LibraryLocator
+from organizations.models import Organization
+
+from . import api
+
+LANGUAGE_TAXONOMY_ID = -1
+ORGANIZATION_TAXONOMY_ID = -3 # -2
+AUTHOR_TAXONOMY_ID = -2 # -3
+
+log = logging.getLogger(__name__)
+
+@shared_task(base=LoggedTask)
```

Comment:
@rpenido One question here, what is the reason why celery tasks are being used?

Reply:
I think this code should run asynchronously (recommended [here](https://docs.openedx.org/projects/openedx-events/en/latest/how-tos/using-events.html#receiving-events)) not to impact the actual update performance (as far as I know, event receivers run synchronously). I don't know yet how costly this operation could be.

Example: If we update an object tag, should we update recursively some children object?

Let me know your thoughts on that!

Reply:
My gut feeling is that celery is overkill for this, and adding a tag to a single component should be extremely fast and lightweight. Do we know for sure that we have to do any recursion over children?

Reply:
> My gut feeling is that celery is overkill for this, and adding a tag to a single component should be extremely fast and lightweight. Do we know for sure that we have to do any recursion over children?

Agree that it is probably overkill, and we probably will not need iterate over children here. 

But isn't it better to leave the "main flow" free from blockers anyway? Don't these little things add up over time? Is the overhead of a celery task call the issue or just code overall complexity?

I'm asking for a broader understanding of how these "secondary" flows are handled in the platform.

Reply:
@rpenido Sorry for the slow reply; I am still trying to figure this out myself.

I actually thought that these events defined in `openedx_events` were always asynchronous but it seems like they are sometimes handled via the "event bus" and sometimes via synchronous Django signals, so I guess I misunderstood something about the architecture.

If you are calling `get_course` in the handler, then it's probably better to do it asynchronously, as that could be a bit slow yeah. But if you changed it to get the language via `CourseOverview` and then save a single tag, that would only represent two database hits and I think that's totally fine in a synchronous handler.

Let's just leave it async for now.

My main concern was just overall code complexity.